File: mobile/lib/ecrans/commande.dart
import 'package:flutter/material.dart';

import '../api/client_api.dart';
import '../charte/jetons.dart';
import '../charte/theme.dart';
import '../modeles/catalogue.dart';
import '../services/services.dart';
import '../widgets/communs.dart';
import 'paiement.dart';

/// Un point de récupération, tel que la route publique le rend.
class PointRecuperation {
  const PointRecuperation({
    required this.id,
    required this.nom,
    required this.ville,
    required this.adresse,
    required this.horaires,
    required this.fraisAcheminement,
  });

  final int id;
  final String nom;
  final String ville;
  final String? adresse;
  final String? horaires;
  final num fraisAcheminement;

  factory PointRecuperation.de(Map<String, dynamic> j) => PointRecuperation(
    id: (j['id'] as num).toInt(),
    nom: (j['nom'] as String?) ?? '',
    ville: (j['ville'] as String?) ?? '',
    adresse: j['adresse'] as String?,
    horaires: j['horaires'] as String?,
    fraisAcheminement: (j['fraisAcheminement'] as num?) ?? 0,
  );
}

class _LignePanier {
  const _LignePanier({
    required this.designation,
    required this.quantite,
    required this.montantLigne,
    required this.vendable,
  });

  final String designation;
  final int quantite;
  final num montantLigne;

  /// Faux si le stock a bougé depuis l'ajout. Le serveur le dit ; l'écran doit
  /// le montrer AVANT le clic, pas après le refus.
  final bool vendable;

  factory _LignePanier.de(Map<String, dynamic> j) => _LignePanier(
    designation: (j['designation'] as String?) ?? 'Article',
    quantite: (j['quantite'] as num?)?.toInt() ?? 0,
    montantLigne: (j['montantLigne'] as num?) ?? 0,
    vendable: (j['vendable'] as bool?) ?? true,
  );
}

class _ContenuPanier {
  const _ContenuPanier({
    required this.lignes,
    required this.montantArticles,
    required this.indisponibles,
  });

  final List<_LignePanier> lignes;
  final num montantArticles;
  final List<String> indisponibles;

  factory _ContenuPanier.de(Map<String, dynamic> j) => _ContenuPanier(
    lignes: (j['lignes'] as List<dynamic>? ?? [])
        .map((e) => _LignePanier.de(e as Map<String, dynamic>))
        .toList(),
    montantArticles: (j['montantArticles'] as num?) ?? 0,
    indisponibles: (j['indisponibles'] as List<dynamic>? ?? [])
        .map((e) => e.toString())
        .toList(),
  );
}

/// Passer commande.
///
/// ## 🎯 UN SEUL ÉCRAN, pas un tunnel
///
/// Il n'y a ni adresse, ni transporteur, ni créneau à choisir : GARAH ne livre
/// pas à domicile. Le seul choix réel est le **lieu de retrait**. Un tunnel en
/// quatre étapes ferait abandonner pour rien.
///
/// ## ⚠️ Le total n'existe qu'une fois le point choisi
///
/// Les frais d'acheminement dépendent du point. Afficher un total avant ce
/// choix serait un chiffre qui change sous les yeux — et un chiffre qui change
/// après qu'on a décidé d'acheter, c'est un achat qu'on ne fait pas.
///
/// ## ⚠️ Le panier vient du SERVEUR, pas du téléphone
///
/// Le panier local sert à ne pas perdre ce qu'on a choisi ; il n'est jamais la
/// vérité. Le stock et le prix se décident côté serveur, et c'est son panier à
/// lui qu'on paie.
class EcranCommande extends StatefulWidget {
  const EcranCommande({super.key});

  @override
  State<EcranCommande> createState() => _EcranCommandeState();
}

class _EcranCommandeState extends State<EcranCommande> {
  List<PointRecuperation> _points = const [];
  _ContenuPanier? _panier;
  int? _pointId;

  bool _chargement = true;
  bool _envoi = false;
  String? _erreur;

  @override
  void initState() {
    super.initState();
    _charger();
  }

  Future<void> _charger() async {
    setState(() {
      _chargement = true;
      _erreur = null;
    });

    final api = Services.de(context).api;
    try {
      final points =
          (await api.obtenir('/api/lieux/points-recuperation') as List<dynamic>)
              .map((e) => PointRecuperation.de(e as Map<String, dynamic>))
              .toList();
      final panier = _ContenuPanier.de(
        await api.obtenir('/api/panier') as Map<String, dynamic>,
      );

      if (!mounted) return;
      setState(() {
        _points = points;
        _panier = panier;
        _chargement = false;
        // ⚠️ AUCUN point pré-choisi. C'est le seul choix de l'écran, et il
        //    décide où l'on ira physiquement chercher sa marchandise : le
        //    pré-remplir ferait valider sans lire, et découvrir la ville au
        //    moment du retrait.
        _pointId = null;
      });
    } on ErreurApi catch (e) {
      if (!mounted) return;
      setState(() {
        _chargement = false;
        _erreur = e.message;
      });
    }
  }

  PointRecuperation? get _point =>
      _points.where((p) => p.id == _pointId).firstOrNull;

  num get _frais => _point?.fraisAcheminement ?? 0;
  num get _total => (_panier?.montantArticles ?? 0) + _frais;

  Future<void> _valider() async {
    if (_pointId == null || _envoi) return;
    setState(() {
      _envoi = true;
      _erreur = null;
    });

    final api = Services.de(context).api;
    final navigateur = Navigator.of(context);
    try {
      final commande =
          await api.poster('/api/commandes', {
                'pointRecuperationId': _pointId,
                'langue': 'fr',
              })
              as Map<String, dynamic>;

      if (!mounted) return;
      setState(() => _envoi = false);
      // On REMPLACE l'écran : revenir en arrière sur un panier déjà commandé
      // ferait passer une seconde commande sans s'en rendre compte.
      navigateur.pushReplacement(
        MaterialPageRoute(
          builder: (_) =>
              EcranPaiement(commandeId: (commande['id'] as num).toInt()),
        ),
      );
    } on ErreurApi catch (e) {
      if (!mounted) return;
      setState(() {
        _envoi = false;
        // Le serveur sait pourquoi il refuse — stock parti, panier vide. Son
        // message est plus juste que celui qu'on inventerait.
        _erreur = e.message;
      });
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text('Passer commande')),
      body: _corps(context),
      bottomNavigationBar: _chargement || _panier == null
          ? null
          : _leTotal(context),
    );
  }

  Widget _corps(BuildContext context) {
    if (_chargement) {
      return const Center(child: CircularProgressIndicator());
    }
    if (_panier == null) {
      return EtatVide(
        message: _erreur ?? 'Votre panier n’a pas pu être chargé.',
        libelleAction: 'Réessayer',
        surAction: _charger,
      );
    }
    if (_panier!.lignes.isEmpty) {
      return const EtatVide(
        message: 'Votre panier est vide.',
        detail: 'Ajoutez des articles avant de commander.',
      );
    }

    return ListView(
      padding: const EdgeInsets.fromLTRB(16, 12, 16, 24),
      children: [
        // ⚠️ Ce qui a disparu du stock se dit EN HAUT, avant même le choix du
        //    point. Le découvrir au moment de valider, c'est refaire tout
        //    l'écran pour rien.
        if (_panier!.indisponibles.isNotEmpty) ...[
          Alerte(
            message:
                'Ces articles ne sont plus disponibles et ne seront pas '
                'commandés :\n${_panier!.indisponibles.join('\n')}',
          ),
          const SizedBox(height: 18),
        ],

        const Libelle('Où retirer votre commande'),
        const SizedBox(height: 4),
        Text(
          'GARAH ne livre pas à domicile : vous venez chercher votre '
          'marchandise au point de votre choix.',
          style: TextStyle(
            fontSize: 12.5,
            height: 1.45,
            color: context.texteAttenue,
          ),
        ),
        const SizedBox(height: 12),
        for (final p in _points) _unPoint(context, p),

        const SizedBox(height: 22),
        const Libelle('Votre commande'),
        const SizedBox(height: 10),
        for (final l in _panier!.lignes) _uneLigne(context, l),

        if (_erreur != null) ...[
          const SizedBox(height: 16),
          Alerte(message: _erreur!),
        ],
      ],
    );
  }

  Widget _unPoint(BuildContext context, PointRecuperation p) {
    final choisi = p.id == _pointId;
    return Container(
      margin: const EdgeInsets.only(bottom: 10),
      decoration: BoxDecoration(
        color: choisi ? Jetons.primaire.withValues(alpha: 0.07) : null,
        border: Border.all(color: choisi ? Jetons.primaire : context.bordure),
        borderRadius: BorderRadius.circular(Jetons.rayonMoyen),
      ),
      child: InkWell(
        onTap: () => setState(() => _pointId = p.id),
        borderRadius: BorderRadius.circular(Jetons.rayonMoyen),
        child: Padding(
          padding: const EdgeInsets.all(14),
          child: Row(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              Icon(
                choisi
                    ? Icons.radio_button_checked
                    : Icons.radio_button_unchecked,
                color: choisi ? Jetons.primaire : context.texteAttenue,
                size: 20,
              ),
              const SizedBox(width: 12),
              Expanded(
                child: Column(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: [
                    Text(
                      p.nom,
                      style: const TextStyle(
                        fontSize: 14.5,
                        fontWeight: FontWeight.w600,
                      ),
                    ),
                    Text(
                      p.ville,
                      style: TextStyle(
                        fontSize: 12.5,
                        color: context.texteAttenue,
                      ),
                    ),
                    if (p.adresse != null && p.adresse!.isNotEmpty)
                      Text(
                        p.adresse!,
                        style: TextStyle(
                          fontSize: 12,
                          color: context.texteAttenue,
                        ),
                      ),
                    if (p.horaires != null && p.horaires!.isNotEmpty)
                      Text(
                        p.horaires!,
                        style: TextStyle(
                          fontSize: 12,
                          color: context.texteAttenue,
                        ),
                      ),
                  ],
                ),
              ),
              const SizedBox(width: 10),
              // 🎯 LES FRAIS SONT SUR LA LIGNE DU POINT, pas dans un
              //    récapitulatif plus bas. Ils font partie du choix : c'est
              //    souvent ce qui départage deux points d'une même ville.
              Column(
                crossAxisAlignment: CrossAxisAlignment.end,
                children: [
                  Text(
                    montantLisible(p.fraisAcheminement),
                    style: const TextStyle(
                      fontSize: 13.5,
                      fontWeight: FontWeight.w700,
                      fontFeatures: [FontFeature.tabularFigures()],
                    ),
                  ),
                  Text(
                    'acheminement',
                    style: TextStyle(
                      fontSize: 10.5,
                      color: context.texteAttenue,
                    ),
                  ),
                ],
              ),
            ],
          ),
        ),
      ),
    );
  }

  Widget _uneLigne(BuildContext context, _LignePanier l) => Padding(
    padding: const EdgeInsets.only(bottom: 8),
    child: Row(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        Expanded(
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              Text(
                l.designation,
                style: TextStyle(
                  fontSize: 13.5,
                  fontWeight: FontWeight.w600,
                  decoration: l.vendable ? null : TextDecoration.lineThrough,
                  color: l.vendable ? null : context.texteAttenue,
                ),
              ),
              Text(
                '${l.quantite} article(s)',
                style: TextStyle(fontSize: 12, color: context.texteAttenue),
              ),
            ],
          ),
        ),
        Text(
          montantLisible(l.montantLigne),
          style: const TextStyle(
            fontSize: 13.5,
            fontWeight: FontWeight.w700,
            fontFeatures: [FontFeature.tabularFigures()],
          ),
        ),
      ],
    ),
  );

  Widget _leTotal(BuildContext context) {
    final choisi = _pointId != null;

    return SafeArea(
      child: Container(
        padding: const EdgeInsets.fromLTRB(16, 12, 16, 12),
        decoration: BoxDecoration(
          color: Theme.of(context).colorScheme.surface,
          border: Border(top: BorderSide(color: context.bordure)),
        ),
        child: Column(
          mainAxisSize: MainAxisSize.min,
          children: [
            _ligneTotal(context, 'Articles', _panier!.montantArticles),
            const SizedBox(height: 4),
            _ligneTotal(
              context,
              'Frais d’acheminement',
              choisi ? _frais : null,
              // Tant qu'aucun point n'est choisi, les frais sont INCONNUS et
              // non nuls. Écrire « 0 FCFA » promettrait la gratuité.
              siInconnu: 'selon le point',
            ),
            const Divider(height: 20),
            Row(
              mainAxisAlignment: MainAxisAlignment.spaceBetween,
              children: [
                const Text(
                  'Total à payer',
                  style: TextStyle(fontSize: 15, fontWeight: FontWeight.w700),
                ),
                Text(
                  choisi ? montantLisible(_total) : '—',
                  style: const TextStyle(
                    fontSize: 19,
                    fontWeight: FontWeight.w700,
                    fontFeatures: [FontFeature.tabularFigures()],
                  ),
                ),
              ],
            ),
            const SizedBox(height: 10),
            // 🎯 Dire ce qui manque AVANT le clic. Un bouton grisé sans raison
            //    fait chercher la panne ailleurs.
            if (!choisi)
              Align(
                alignment: Alignment.centerLeft,
                child: Text(
                  'Choisissez le point où vous viendrez retirer votre commande.',
                  style: const TextStyle(fontSize: 12.5, color: Jetons.alerte),
                ),
              ),
            if (!choisi) const SizedBox(height: 8),
            SizedBox(
              width: double.infinity,
              child: FilledButton(
                onPressed: choisi && !_envoi && _panier!.lignes.isNotEmpty
                    ? _valider
                    : null,
                child: Text(_envoi ? 'Enregistrement…' : 'Commander et payer'),
              ),
            ),
          ],
        ),
      ),
    );
  }

  Widget _ligneTotal(
    BuildContext context,
    String libelle,
    num? valeur, {
    String? siInconnu,
  }) => Row(
    mainAxisAlignment: MainAxisAlignment.spaceBetween,
    children: [
      Text(
        libelle,
        style: TextStyle(fontSize: 13, color: context.texteAttenue),
      ),
      Text(
        valeur != null ? montantLisible(valeur) : (siInconnu ?? '—'),
        style: TextStyle(
          fontSize: 13.5,
          fontWeight: FontWeight.w600,
          color: valeur == null ? context.texteAttenue : null,
          fontFeatures: const [FontFeature.tabularFigures()],
        ),
      ),
    ],
  );
}

File: web/src/app/commande/commande.ts
import { HttpClient } from '@angular/common/http';
import { Component, computed, inject, signal } from '@angular/core';
import { Router, RouterLink } from '@angular/router';
import { catchError, of } from 'rxjs';

import { montantLisible } from '../../modeles/catalogue';
import { messageErreur } from '../../api/erreurs';
import { PanierLocal } from '../../services/panier-local';
import { ServiceSession } from '../../services/session';

/** Un point de récupération, tel que la route publique le rend. */
interface PointRecuperation {
  readonly id: number;
  readonly nom: string;
  readonly ville: string;
  readonly pays: string;
  readonly adresse: string | null;
  readonly horaires: string | null;
  readonly fraisAcheminement: number;
}

interface LignePanier {
  readonly varianteId: number;
  readonly designation: string;
  readonly quantite: number;
  readonly prixUnitaire: number;
  readonly montantLigne: number;
  readonly vendable: boolean;
  /*
   * Vrai si le prix vient d'une negociation acceptee dans une discussion.
   * Le dire : sinon le client verrait un prix different de la fiche produit
   * et croirait a une erreur — dans un sens ou dans l'autre.
   */
  readonly prixNegocie?: boolean;
}

interface ContenuPanier {
  readonly lignes: readonly LignePanier[];
  readonly montantArticles: number;
  readonly nombreArticles: number;
  readonly indisponibles: readonly string[];
}

/**
 * Passer commande.
 *
 * <h2>🎯 UN SEUL ÉCRAN, pas un tunnel</h2>
 *
 * <p>Il n'y a ni adresse, ni transporteur, ni créneau à choisir : GARAH ne
 * livre pas à domicile (D-05). Le seul choix réel est le <b>lieu de
 * retrait</b> — et il change le total, puisqu'il porte les frais
 * d'acheminement.</p>
 *
 * <p>Un tunnel en quatre étapes ferait abandonner pour rien.</p>
 *
 * <p>⚠️ Le panier lu ici est celui du <b>serveur</b>, pas le panier local :
 * lui seul connaît le stock, les paliers de quantité et le prix du jour.</p>
 */
@Component({
  selector: 'gb-commande',
  imports: [RouterLink],
  template: `
    <header class="entete">
      <h1>Passer la commande</h1>
      <p class="entete__aide">Un seul choix : où retirer.</p>
    </header>

    @if (erreur(); as m) {
      <p class="gb-alerte marge" role="alert">{{ m }}</p>
    }

    @if (chargement()) {
      <div class="gb-etat"><p>Chargement…</p></div>
    } @else {

      <!-- Ce que le serveur refuse de vendre. À dire AVANT le bouton : le
           découvrir après le clic ferait revenir en arrière. -->
      @if (panier()?.indisponibles?.length) {
        <p class="gb-alerte marge" role="alert">
          Certains articles ne sont plus disponibles et ne seront pas commandés :
          {{ panier()!.indisponibles.join(', ') }}
        </p>
      }

      <section class="bloc">
        <p class="gb-libelle">Point de récupération</p>

        @for (p of points(); track p.id) {
          <button type="button" class="point" [class.point--choisi]="p.id === pointId()"
                  (click)="pointId.set(p.id)">
            <span class="point__marque" aria-hidden="true"></span>
            <span class="point__corps">
              <span class="point__tete">
                <span class="point__nom">{{ p.nom }}</span>
                <span class="point__frais">{{ montant(p.fraisAcheminement) }}</span>
              </span>
              @if (p.adresse) { <span class="point__detail">{{ p.adresse }}</span> }
              @if (p.horaires) { <span class="point__detail">{{ p.horaires }}</span> }
            </span>
          </button>
        }

        <!-- Ni adresse, ni livraison. Le dire évite de chercher le champ. -->
        <p class="bloc__aide">
          GARAH ne livre pas à domicile. Vous retirez vous-même, avec un code
          qu’on vous remettra.
        </p>
      </section>

      @if (panier(); as c) {
        <section class="gb-carte recap">
          <p class="gb-libelle">Votre commande</p>

          @for (l of c.lignes; track l.varianteId) {
            <div class="recap__ligne">
              <span>
                {{ l.designation }} <span class="gb-attenue">× {{ l.quantite }}</span>
                @if (l.prixNegocie) { <span class="gb-attenue">· prix négocié</span> }
              </span>
              <span class="gb-montant">{{ montant(l.montantLigne) }}</span>
            </div>
          }

          <div class="recap__ligne">
            <span class="gb-attenue">Acheminement</span>
            <span class="gb-montant">{{ montant(frais()) }}</span>
          </div>

          <div class="recap__separateur"></div>

          <div class="recap__ligne recap__ligne--total">
            <span>À payer</span>
            <span class="gb-montant recap__total">{{ montant(total()) }}</span>
          </div>

          <!-- Les montants sont figés ICI. Le dire avant, pas après. -->
          <p class="recap__mention">
            Ces montants sont figés à la validation. Ils ne bougeront plus, même
            si les tarifs changent.
          </p>
        </section>

        <div class="actions">
          <button type="button" class="gb-btn gb-btn--primaire gb-btn--plein"
                  [disabled]="pointId() === null || envoi() || c.lignes.length === 0"
                  (click)="valider()">
            @if (envoi()) { Validation… } @else { Valider et payer }
          </button>
          <!-- On ne s'annule pas soi-même après paiement (D-12) : le dire AVANT
               de payer, pas au moment de le découvrir. -->
          <p class="actions__mention">
            Une fois payée, une commande s’annule en nous écrivant, pas d’un clic.
          </p>
        </div>
      } @else {
        <div class="gb-etat">
          <p>Votre panier est vide.</p>
          <a routerLink="/catalogue" class="gb-btn gb-btn--secondaire">Parcourir le catalogue</a>
        </div>
      }
    }
  `,
  styles: `
    .entete { padding: 1.5rem 1.25rem 1rem; }
    .entete h1 { margin: 0; font-size: 1.3rem; font-weight: 700; }
    .entete__aide { margin: 0.3rem 0 0; font-size: 0.82rem; color: var(--texte-attenue); }

    .marge { margin: 0 1.25rem 1rem; }

    .bloc { padding: 0 1.25rem; display: flex; flex-direction: column; gap: 0.6rem; }
    .bloc__aide { margin: 0.3rem 0 0; font-size: 0.78rem; color: var(--texte-attenue); line-height: 1.55; }

    .point {
      display: flex;
      align-items: flex-start;
      gap: 0.75rem;
      width: 100%;
      min-height: var(--cible-tactile-min);
      padding: 0.9rem;
      border-radius: var(--rayon-moyen);
      border: 1px solid var(--verre-bordure);
      background: var(--verre-fond);
      color: inherit;
      font-family: inherit;
      text-align: left;
      cursor: pointer;

      &:focus-visible { outline: 2px solid var(--primaire); outline-offset: 2px; }
    }

    .point--choisi {
      border: 2px solid var(--primaire);
      background: color-mix(in srgb, var(--primaire) 6%, transparent);

      .point__marque { border: 6px solid var(--primaire); }
    }

    .point__marque {
      width: 20px; height: 20px; flex-shrink: 0; margin-top: 0.1rem;
      border-radius: 50%;
      border: 1px solid var(--champ-bordure);
      background: var(--surface);
    }

    .point__corps { flex: 1; min-width: 0; display: flex; flex-direction: column; gap: 0.2rem; }

    .point__tete { display: flex; align-items: baseline; justify-content: space-between; gap: 0.6rem; }
    .point__nom { font-size: 0.9rem; font-weight: 600; }
    .point__frais { font-size: 0.85rem; font-weight: 700; white-space: nowrap; font-variant-numeric: tabular-nums; }
    .point__detail { font-size: 0.78rem; color: var(--texte-attenue); line-height: 1.45; }

    .recap { margin: 1.25rem 1.25rem 0; padding: 1rem; display: flex; flex-direction: column; gap: 0.55rem; }
    .recap__ligne { display: flex; align-items: baseline; justify-content: space-between; gap: 0.75rem; font-size: 0.85rem; }
    .recap__ligne--total { font-size: 1rem; font-weight: 700; }
    .recap__total { font-size: 1.3rem; }
    .recap__separateur { height: 1px; background: var(--verre-bordure); margin: 0.2rem 0; }
    .recap__mention { margin: 0.3rem 0 0; font-size: 0.75rem; color: var(--texte-attenue); line-height: 1.55; }

    .actions { padding: 1.25rem; display: flex; flex-direction: column; gap: 0.6rem; }
    .actions__mention { margin: 0; text-align: center; font-size: 0.75rem; color: var(--texte-attenue); line-height: 1.5; }
  `,
})
export class Commande {
  private readonly http = inject(HttpClient);
  private readonly router = inject(Router);
  private readonly session = inject(ServiceSession);
  private readonly panierLocal = inject(PanierLocal);

  protected readonly points = signal<readonly PointRecuperation[]>([]);
  protected readonly panier = signal<ContenuPanier | null>(null);
  protected readonly pointId = signal<number | null>(null);

  protected readonly chargement = signal(true);
  protected readonly envoi = signal(false);
  protected readonly erreur = signal<string | null>(null);

  protected readonly frais = computed(
    () => this.points().find((p) => p.id === this.pointId())?.fraisAcheminement ?? 0,
  );

  protected readonly total = computed(
    () => (this.panier()?.montantArticles ?? 0) + this.frais(),
  );

  constructor() {
    this.charger();
  }

  /**
   * Les points de récupération et le panier.
   *
   * <p>⚠️ LE PANIER LOCAL EST POUSSÉ AU SERVEUR AVANT D'ÊTRE RELU. L'écran lit
   * {@code /api/panier} — le panier du serveur — alors que l'écran précédent
   * affiche celui du navigateur. Sans cette synchronisation, tout ce qu'un
   * client déjà connecté a ajouté restait dans son navigateur, et cet écran
   * annonçait « Votre panier est vide » sur un panier qui affichait ses lignes
   * et son montant.</p>
   *
   * <p>La synchronisation est faite ICI et non sur le bouton de l'écran
   * précédent : on arrive aussi sur {@code /commande} par l'URL, par le retour
   * arrière, ou après une connexion qui redirige. Un seul de ces chemins passe
   * par le bouton.</p>
   *
   * <p>⚠️ Son échec n'empêche PAS de continuer. Le serveur a peut-être déjà le
   * panier — d'une session précédente, d'un autre appareil. S'arrêter là
   * priverait d'une commande possible ; on relit donc le panier serveur et
   * c'est lui qui dira s'il est vide.</p>
   */
  private charger(): void {
    this.chargement.set(true);

    this.http.get<PointRecuperation[]>('/api/lieux/points-recuperation').subscribe({
      next: (p) => {
        this.points.set(p);
        // Le moins cher d'abord : c'est le choix par défaut le plus sûr, et
        // celui qu'on regrette le moins si on ne regarde pas.
        const moinsCher = [...p].sort((a, b) => a.fraisAcheminement - b.fraisAcheminement)[0];
        this.pointId.set(moinsCher?.id ?? null);
      },
      error: () => this.erreur.set('Les points de récupération n’ont pas pu être chargés.'),
    });

    this.session
      .synchroniserLePanier()
      .pipe(catchError(() => of(null)))
      .subscribe(() => this.lirePanier());
  }

  private lirePanier(): void {
    this.http.get<ContenuPanier>('/api/panier').subscribe({
      next: (c) => {
        this.panier.set(c.lignes.length > 0 ? c : null);
        this.chargement.set(false);
      },
      error: () => {
        this.chargement.set(false);
        this.erreur.set('Votre panier n’a pas pu être chargé.');
      },
    });
  }

  protected valider(): void {
    const point = this.pointId();
    if (point === null || this.envoi()) {
      return;
    }
    this.envoi.set(true);
    this.erreur.set(null);

    this.http
      .post<{ id: number }>('/api/commandes', { pointRecuperationId: point, langue: 'fr' })
      .subscribe({
        next: (c) => {
          this.envoi.set(false);
          // ⚠️ LE PANIER LOCAL SE VIDE ICI, et pas avant.
          //
          //    La commande est passée : le serveur a consommé SON panier, et
          //    le local n'a plus de raison d'être. Le vider plus tôt — à la
          //    synchronisation, comme le faisait l'ancienne fusion — vidait
          //    l'écran du panier alors qu'aucune commande n'était passée.
          this.panierLocal.vider();
          void this.router.navigate(['/paiement', c.id]);
        },
        error: (e: unknown) => {
          this.envoi.set(false);
          this.erreur.set(messageErreur(e, 'La commande n’a pas pu être créée.'));
        },
      });
  }

  protected montant(valeur: number): string {
    return montantLisible(valeur);
  }
}

